### PR TITLE
Fogl 1523

### PR DESCRIPTION
**Pull request by billehunt**
changes to enable jq transformations in sinusoid plugin. See Jira ticket for samples to test with.

### CODE REVIEW

**Review by MarkRiddoch — COMMENTED** [review]
This functionality would be better placed in the foglamp south service. That way it can be applied to any plugin, not just the sinusoid.

Plugins should be the minimal unique functionality needed to interact with a particular device, functionality that is not related to a particular device or is useful for more than one device should really live in the service to prevent duplication.

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 1]
```diff
@@ -168,12 +183,29 @@ def generate_data():
                         "sinusoid": next(generate_data())
                     }
                 }
+                # _LOGGER.info("Original Sinusoid data: {}".format(data))
+
+                # Handles the JQFilter functionality
+                if handle['applyFilter']["value"].upper() == "TRUE":
+
+                    # Steps needed to proper format the data generated by the JQFilter
+                    # to the one expected by the SP
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py:192]: SP ?

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 2]
```diff
@@ -168,12 +183,29 @@ def generate_data():
...
+                    data_2 = jqfilter.transform(data, handle['filterRule']["value"])
+                    data_3 = json.dumps(data_2)
+                    del data_2
+
+                    data_4 = eval(data_3)
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py:197]: ast.literal_eval is safe choice and an exception handler should be added corresponding to that

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 3]
```diff
@@ -168,12 +183,29 @@ def generate_data():
...
+                    del data_3
+
+                    data = data_4[0]
+                    del data_4
+                    # _LOGGER.info("Transformed Sinusoid data: {}".format(data))
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py:202]: could we please make variable names better; I believe we can avoid data_2, _3, _4 with a better expression;

File: python/foglamp/plugins/south/sinusoid/sinusoid.py [hunk 4]
```diff
@@ -199,15 +231,10 @@ def plugin_reconfigure(handle, new_config):
...
-        new_handle = plugin_init(new_config)
-        new_handle['restart'] = 'yes'
-        _LOGGER.info("Restarting Sinusoid plugin due to change in configuration key [{}]".format(', '.join(diff)))
-    else:
-        new_handle = copy.deepcopy(handle)
-        new_handle['restart'] = 'no'
+    _plugin_stop(handle)
+    new_handle = plugin_init(new_config)
+    new_handle['restart'] = 'yes'
+    _LOGGER.info("Restarting Sinusoid plugin due to change in configuration key [{}]".format(', '.join(diff)))
```
Comment (ashish-jabble) [inline python/foglamp/plugins/south/sinusoid/sinusoid.py:237]: do we want to restart always? instead of specific config item's value change!

**Review by ashish-jabble — CHANGES_REQUESTED** [review]

File: python/foglamp/plugins/south/sinusoid/sinusoid.py (hunk 1 above)
Comment (billehunt) [reply python/foglamp/plugins/south/sinusoid/sinusoid.py:192]: This code was directly copied from foglamp/tasks/north/sending_process.py. I'm not sure what SP is, and I agree: this comment, and the names data_2, data_3, data_4 can be improved.

File: python/foglamp/plugins/south/sinusoid/sinusoid.py (hunk 4 above)
Comment (billehunt) [reply python/foglamp/plugins/south/sinusoid/sinusoid.py:237]: I'm going to pull the code and revert to the previous only restarting when necessary. However, the previous only prevented restarting if the plugin name was changed, which should never happen, so I think practically this isn't a concern.

**Comment by billehunt** [comment]
I agree with Mark, that this functionality should go into the south service. Until we have it there, I'll keep this code in a separate branch for demo purposes, and once it's in south service, will kill it.

**Pull request closed without merging**